sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/TransientIOErrorsRetryingIterator.scala

```diff
@@ -36,37 +39,64 @@ private class TransientIOErrorsRetryingIterator
 
   private[spark] var maxRetryIntervalInMs = CosmosConstants.maxRetryIntervalForTransientFailuresInMs
   private[spark] var maxRetryCount = CosmosConstants.maxRetryCountForTransientFailures
-  private val lock = new ReentrantReadWriteLock()
-  private val read = lock.readLock()
-  private val write = lock.writeLock()
 
   private val rnd = Random
   // scalastyle:off null
   private val lastContinuationToken = new AtomicReference[String](null)
   // scalastyle:on null
   private val retryCount = new AtomicLong(0)
-  private var currentCosmosPagedIterable: CosmosPagedIterable[ObjectNode] = _
-  private[spark] var currentIterator: java.util.Iterator[ObjectNode] = _
 
-  reinitialize()
+  private[spark] var currentFeedResponseIterator: Option[Iterator[FeedResponse[ObjectNode]]] = None
+  private[spark] var currentItemIterator: Option[Iterator[ObjectNode]] = None
 
   override def hasNext: Boolean = {
-    executeWithRetry("hasNext", () => currentIterator.hasNext)
+    executeWithRetry("hasNextInternal", () => hasNextInternal)
   }
 
-  override def next(): ObjectNode = {
-    executeWithRetry("next", () => currentIterator.next)
+  private def hasNextInternal: Boolean = {
+    if (hasBufferedNext) {
+      true
+    } else {
+      val feedResponseIterator = currentFeedResponseIterator match {
+        case Some(existing) => existing
+        case None => {
+          currentFeedResponseIterator = Some(new CosmosPagedIterable[ObjectNode](
```

Comment:
I thought we were using `CosmosPagedFlux`. This is all synchronous, which is fine if it fits the use case, just surprised to see it as you mentioned about using `CosmosPagedFlux`

Reply:
Spark's API is sync - just mentioned CosmsoPagedFlux because that is what is passed in - but conversion to PagedIterable happened even before - and aligns with usage in the Spark API